subcommand/mesh-init/command_test.go

```diff
@@ -348,6 +451,128 @@ func TestRun(t *testing.T) {
+			// of syncChecks
+			var taskMetaContainersResponse string
+			if !c.missingDataplaneContainer {
+				taskMetaContainersResponse = fmt.Sprintf(`{"Name": "%s", "Health": {"Status": "%s"}}`, config.ConsulDataplaneContainerName, ecs.HealthStatusHealthy)
+			}
+			for name, hsc := range c.healthSyncContainers {
+				if hsc.missing {
+					continue
+				}
+
+				if taskMetaContainersResponse != "" {
+					taskMetaContainersResponse += ","
+				}
+
+				taskMetaContainersResponse += fmt.Sprintf(`{"Name": "%s", "Health": {"Status": "%s"}}`, name, hsc.status)
+			}
+			taskMetadataResponse = fmt.Sprintf(`{"Cluster": "test", "TaskARN": "%s", "Family": "%s", "Containers": [%s]}`, taskARN, family, taskMetaContainersResponse)
+			testutil.TaskMetaServer(t, testutil.TaskMetaHandler(t, taskMetadataResponse))
+
+			// Populate expectedServiceChecks based on the new status
+			// of containers returned by the task meta server
+			for _, expCheck := range expectedServiceChecks {
+				found := false
+				if expCheck.Name == consulDataplaneReadinessCheckName && c.missingDataplaneContainer {
+					expCheck.Status = api.HealthCritical
+					continue
+				}
+
+				for name, hcs := range c.healthSyncContainers {
+					checkID := constructCheckID(expServiceID, name)
+					if expCheck.CheckID == checkID {
+						if hcs.missing {
+							expCheck.Status = api.HealthCritical
+						} else {
+							expCheck.Status = ecsHealthToConsulHealth(hcs.status)
+						}
+						found = true
+						break
+					}
+				}
+
+				if !found {
+					expCheck.Status = api.HealthPassing
+				}
+			}
+
+			expectedProxyCheck.Status = api.HealthPassing
+			if c.missingDataplaneContainer {
+				expectedProxyCheck.Status = api.HealthCritical
+			}
+
+			// Signals control plane to enter into a state where it
+			// periodically sync checks back to Consul
+			close(cmd.proceedChan)
+
+			// Verify with retries that the checks have reached the expected state
+			assertHealthChecks(t, consulClient, expectedServiceChecks, expectedProxyCheck)
+
+			// Some containers might reappear after sometime they went missing.
+			// This block makes a missing reappear in the task meta response and
+			// tests if the control plane is able to sync back the status of the
+			// container to Consul servers.
+			if c.shouldMissingContainersReappear {
+				err = os.Unsetenv(awsutil.ECSMetadataURIEnvVar)
+				require.NoError(t, err)
+				taskMetaContainersResponse = fmt.Sprintf(`{"Name": "%s", "Health": {"Status": "%s"}}`, config.ConsulDataplaneContainerName, ecs.HealthStatusHealthy)
+				for name, hsc := range c.healthSyncContainers {
+					taskMetaContainersResponse += fmt.Sprintf(`,{"Name": "%s", "Health": {"Status": "%s"}}`, name, hsc.status)
+				}
+				taskMetadataResponse = fmt.Sprintf(`{"Cluster": "test", "TaskARN": "%s", "Family": "%s", "Containers": [%s]}`, taskARN, family, taskMetaContainersResponse)
+				testutil.TaskMetaServer(t, testutil.TaskMetaHandler(t, taskMetadataResponse))
```

Comment:
This is the second TaskMetaServer started in this test, which I guess is why we need to unset the `ECSMetadataURIEnvVar` in this case.

Instead of starting two servers, you could have the handler conditionally return different responses at different points of the test. Something simple like:

```go
var currentTaskMetaResp atomic.Value
currentTaskMetaResp.Store("<taskMetaJSON>")
testutil.TaskMetaServer(t, testutil.TaskMetaHandlerFn(t,
	func() string {
		return currentTaskMetaResp.Load().(string)
	},
))
```

Then you can call `currentTaskMetaResp.Store("<taskMetaJSON>")` whenever you want it to return a new response. (And we could make a helper type that wraps this up in the `testutil` package)

Reply:
Thanks. Modified this